src/lib.rs

```diff
@@ -1,20 +1,118 @@
 mod error;
-mod graph;
-mod loader;
-mod parser;
-mod reqwest_loader;
-mod resolve_import;
-
-pub use error::Error;
-pub use graph::ModuleGraph;
-pub use graph::GRAPH_VERSION;
-pub use loader::MemoryLoader;
-pub use loader::ModuleInfo;
-pub use loader::ModuleLoad;
-pub use loader::ModuleLoadFuture;
-pub use loader::ModuleLoader;
-pub use loader::ModuleSource;
-pub use loader::ModuleStream;
-pub use reqwest_loader::load_reqwest;
-pub use reqwest_loader::none_middleware;
-pub use reqwest_loader::ReqwestLoader;
+pub mod v1;
+pub mod v2;
+
+use std::pin::Pin;
+use std::sync::Arc;
+
+use futures::Future;
+use serde::Deserialize;
+use serde::Serialize;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::AsyncReadExt;
+use v2::ESZIP_V2_MAGIC;
+
+pub use crate::error::ParseError;
+pub use crate::v1::EsZipV1;
+pub use crate::v2::EsZipV2;
+
+pub use deno_graph;
+
+pub enum EsZip {
+  V1(EsZipV1),
+  V2(EsZipV2),
+}
+
+impl EsZip {
+  pub async fn parse<R: tokio::io::AsyncRead + Unpin + 'static>(
+    reader: R,
+  ) -> Result<
+    (EsZip, Pin<Box<dyn Future<Output = Result<(), ParseError>>>>),
```

Comment:
Consider factoring `Pin<Box<dyn Future<Output = Result<(), ParseError>>>>` to a separate type with a docstring explaining that the future needs to be driven forward for `EsZip` to return values

Reply:
Done